### PR TITLE
refactor: Remove private frameworks from the list of imports

### PR DESCRIPTION
**Pull request by mykola-mokhnach**
This PR removes the explicit dependency to XCTAutomationSupport private framework because of the resends described in https://github.com/appium/appium/issues/17174

The removal of  the framework forces us to also remove all dependant classes, including XCElementSnapshot one. This PR replaces these dependant classes with interfaces. The only issue that ObjC does not allow to use default method implementations in protocol definitions, so we either need to introduce global helper methods for custom properties/extension methods or have a custom wrapper over XCElementSnapshot instances each time we need to call a customized entity on it. I've selected the second approach.

### CODE REVIEW

**Comment by mykola-mokhnach** [comment]
it looks like the most complicated part there would be to make NSPredicate work properly with custom attributes. I knew it was going to explode some day.

Basically we need to parse the initial predicate expression as a string and transform it into a block, where we could process custom snapshot properties (e.g. these starting with wd...)

**Comment by Dan-Maor** [comment]
@mykola-mokhnach we can instead use `class_addMethod` in runtime for `XCElementSnapshot` and add our attributes.
I've tested it in a quick and dirty fashion by exposing `wdLabel` and `isWDVisible` from `XCUIElement+FBWebDriverAttributes` and added the following code to `FBXCElementSnapshotWrapper`:

```
+ (void)load
{
  Class XCElementSnapshotCls = objc_lookUpClass("XCElementSnapshot");
  
  // for wdLabel
  Method wdLabelMethod = class_getInstanceMethod([FBXCElementSnapshotWrapper class], @selector(wdLabel));
  class_addMethod(XCElementSnapshotCls, @selector(wdLabel), method_getImplementation(wdLabelMethod), method_getTypeEncoding(wdLabelMethod));
  
  // for isWDVisible
  Method isWdVisibleMethod = class_getInstanceMethod([FBXCElementSnapshotWrapper class], @selector(isWDVisible));
  class_addMethod(XCElementSnapshotCls, @selector(isWDVisible), method_getImplementation(isWdVisibleMethod), method_getTypeEncoding(isWdVisibleMethod));
}
```

For `isWDVisible` I also had to modify it to the following (since `fb_isVisible` is not defined for `XCElementSnapshot`, but on second thought we can also add this method directly as well as above):
```
- (BOOL)isWDVisible
{
  return [[FBXCElementSnapshotWrapper alloc] initWithSnapshot:self].fb_isVisible;
}
```

File: WebDriverAgentLib/Routing/FBXCDeviceEvent.m [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+#import "FBXCDeviceEvent.h"
+
+#import "FBErrorBuilder.h"
+
+id<FBXCDeviceEvent> FBCreateXCDeviceEvent(unsigned int page,
+                                          unsigned int usage,
+                                          double duration,
+                                          NSError **error)
+{
+  Class xcDeviceEventClass = NSClassFromString(@"XCDeviceEvent");
```
Comment (Dan-Maor) [inline WebDriverAgentLib/Routing/FBXCDeviceEvent.m:19]: Perhaps we should use dispatch_once here?

File: WebDriverAgentLib/Utilities/XCTestPrivateSymbols.m [hunk 2]
```diff
@@ -56,8 +55,9 @@
 
 NSArray<NSString*> *FBStandardAttributeNames(void)
 {
-  return [XCElementSnapshot sanitizedElementSnapshotHierarchyAttributesForAttributes:nil
-                                                                             isMacOS:NO];
+  Class xcElementSnapshotClass = NSClassFromString(@"XCElementSnapshot");
```
Comment (Dan-Maor) [inline WebDriverAgentLib/Utilities/XCTestPrivateSymbols.m]: Perhaps we should use dispatch_once here as well?

File: WebDriverAgentLib/Routing/FBXCDeviceEvent.m (hunk 1 above)
Comment (mykola-mokhnach) [reply WebDriverAgentLib/Routing/FBXCDeviceEvent.m:19]: is it so expensive to create class from string?

File: WebDriverAgentLib/Utilities/XCTestPrivateSymbols.m (hunk 2 above)
Comment (mykola-mokhnach) [reply WebDriverAgentLib/Utilities/XCTestPrivateSymbols.m]: will do

File: WebDriverAgentLib/Routing/FBXCDeviceEvent.m (hunk 1 above)
Comment (Dan-Maor) [reply WebDriverAgentLib/Routing/FBXCDeviceEvent.m:19]: Probably not much, just a matter of style really. I think this instance can be left as is.

**Review by Dan-Maor — APPROVED** [review]

File: Gemfile [hunk 3]
```diff
@@ -2,6 +2,6 @@ source "https://rubygems.org"
 
 ruby ">= 3"
 
-gem "fastlane"
+gem "fastlane", "2.201.1"
```
Comment (KazuCocoa) [inline Gemfile]: The fix has not been merged, so maybe the below will help. 
`gem "fastlane", git: git@github.com:kentrh/fastlane.git, branch: bug/20012-scan-cannot-find-xcresult-testing-swift-package`

Comment (KazuCocoa) [reply Gemfile]: ah, or if this branch did not work, we can fork (e.g. to my repository) the fastlane and create a branch with the fix. Then, this gem can refer to it.

Comment (KazuCocoa) [reply Gemfile]: okay, i've created a branch with the fix
`gem "fastlane", git: git@github.com:KazuCocoa/fastlane.git, branch: scan-swift-package`

**Review by KazuCocoa — APPROVED** [review]
lg. I've checked a few with a real device

File: Gemfile [hunk 4]
```diff
@@ -2,6 +2,6 @@ source "https://rubygems.org"
 
 ruby ">= 3"
 
-gem "fastlane"
+gem "fastlane", git: "git@github.com:kentrh/fastlane.git", branch: "bug/20012-scan-cannot-find-xcresult-testing-swift-package"
```
Comment (KazuCocoa) [inline Gemfile]: what about `https://github.com/kentrh/fastlane.git` instead of `git@github.com:kentrh/fastlane.git`?

**Comment by KazuCocoa** [comment]
What about the below fastlane version?
https://dev.azure.com/AppiumCI/Appium%20CI/_build/results?buildId=20298&view=logs&j=b0b7e611-f24f-5025-cfe1-d0e5eed5ba7f&t=1db81a0e-a352-5921-8e54-3d6dffa5f477

```
Fetching fastlane 2.162.0
Installing fastlane 2.162.0
```

**Comment by KazuCocoa** [comment]
let me try 2.162.0...

**Comment by mykola-mokhnach** [comment]
Thanks @KazuCocoa 
I'd say we are happy with this version for now + Ruby2 and stay with it until possible.

**Comment by timofeys1** [comment]
One way we found to link with XCTAutomationSupport was to use a fake dylib with a name from the allowed names and reexport XCTAutomationSupport through `-reexport_framework`/`-reexport_library` linker flag. I made an [example here](https://github.com/abdulowork/LinkWithXCTAutomationSupportXcode14). 

A bit of a hack but maybe that's something that you would want to employ?

**Comment by pwfcurry** [comment]
Thanks for the fix @mykola-mokhnach!

Considering appium 2 is in beta, are there any plans to fix this for appium 1? (it still pulls in appium-webdriveragent v3, fix is in v4)

(couple of workarounds: use your old `~/Library/Developer/Xcode/DerivedData/WebDriverAgent-...` dir built with xcode 13 (rename to match expected), or pin v4 of appium-webdriveragent)

**Comment by jlipps** [comment]
@pwfcurry no, appium 1 is now unsupported. appium 2 is only in beta because the documentation hasn't been ready for a release. but in terms of its suitability for production use code-wise, it has been our only supported version for some time now. so it would make sense to update to appium 2 to take advantage of any fixes.

**Pull request merged**